### ISSUE TITLE
The value of the 'content' field

### ISSUE THREAD

**Issue by 4kb-wgh** [elsa-workflows/elsa-core#1534]
`
{
  "activityId": "cd3a8a78-d3c3-43ee-8b3e-5685c41810d8",
  "displayName": "Send HTTP Request",
  "name": "resName",
  "type": "SendHttpRequest",
  "properties": [
    {
      "name": "Url",
      "value": "https://partnerdev-******/q"
    },
    {
      "name": "Method",
      "value": "POST"
    },
    {
      "name": "Content",
      "value": "System.Dynamic.ExpandoObject"
    },
    {
      "name": "ContentType",
      "value": "application/json"
    },
    {
      "name": "Authorization",
      "value": null
    },
    {
      "name": "RequestHeaders",
      "value": {}
    },
    {
      "name": "ReadContent",
      "value": true
    },
    {
      "name": "SupportedStatusCodes",
      "value": [
        200
      ]
    }
  ],
  "outcomes": [
    "Done"
  ],
  "persistWorkflow": false,
  "saveWorkflowContext": false,
  "loadWorkflowContext": false
}
`

The value of the 'content' field is 'System.Dynamic.ExpandoObject'. 
why? ? ?
Then I find the source code
` [ActivityInput(Hint = "The HTTP content to send along with the request.", UIHint = ActivityInputUIHints.MultiLine, SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid })]
`
`public string? Content { get; set; }`
So
How do I convert strings to objects?

**Comment by 4kb-wgh** [comment]
@sfmskywalker Help me see this problem. Thanks

**Comment by 4kb-wgh** [comment]
![image](https://user-images.githubusercontent.com/63579509/133578429-22952652-7042-4e2e-9233-9525494aac1b.png)

![image](https://user-images.githubusercontent.com/63579509/133578834-c2ce0608-9553-4f26-855b-36d8a1f754ca.png)

![image](https://user-images.githubusercontent.com/63579509/133578947-f5679083-dbd2-4231-86b1-dbeac75f36a6.png)

**Comment by sfmskywalker** [comment]
@4kb-wgh 

The `Content` to send to the HTTP endpoint needs to be a string :)

So in your case, just send your object as a JSON string, like so:

```json
{
   "TableName": "v_get_sysdic_list",
   "Where": "fatherid=22"
}
```

**Comment by 4kb-wgh** [comment]
I want to send to HTTP request with object, but the `Content` is a string.
When i send to HTTP request with a string, it throw exception
`Exception
System.ArgumentException
The value "" is not of type "System.Dynamic.ExpandoObject" and cannot be used in this generic collection. (Parameter 'value')`

**Comment by 4kb-wgh** [comment]
> @4kb-wgh
> 
> The `Content` to send to the HTTP endpoint needs to be a string :)
> 
> So in your case, just send your object as a JSON string, like so:
> 
> ```json
> {
>    "TableName": "v_get_sysdic_list",
>    "Where": "fatherid=22"
> }
> ```

I want to send to HTTP request with object, but the `Content` is a string.
When i send to HTTP request with a string, it throw exception
`Exception
System.ArgumentException
The value "" is not of type "System.Dynamic.ExpandoObject" and cannot be used in this generic collection. (Parameter 'value')`

**Comment by sfmskywalker** [comment]
I just tried and it works fine for me. If you send me sample project demonstrating the issue then I can investigate if you want. Or you could try to use the latest Elsa 2.3 preview packages available on MyGet to see if you are still getting this issue (if so, I definitely need a repro :) )

**Comment by 4kb-wgh** [comment]
My project is Elas 2.2.1 in NuGet.
![image](https://user-images.githubusercontent.com/63579509/133696564-e445dd9e-761a-4087-9c26-85421477d60a.png)

How can i use the latest Elsa 2.3 preview packages

> I just tried and it works fine for me. If you send me sample project demonstrating the issue then I can investigate if you want. Or you could try to use the latest Elsa 2.3 preview packages available on MyGet to see if you are still getting this issue (if so, I definitely need a repro 😄)

**Comment by 4kb-wgh** [comment]
Can the `Content` be an object or a JObject?
![image](https://user-images.githubusercontent.com/63579509/133702628-0f501a24-4e59-413a-9348-7eb17a679429.png)

**Comment by 4kb-wgh** [comment]
Sorry, the exception is not caused by the question above. 
It's caused by the return value of my interface, like so:
Object[] objectA = new Object[2];
objectA[0] = "";
objectA[1] = "SUCCESS";
return objectA;

![image](https://user-images.githubusercontent.com/63579509/133736286-9dc42e15-aad2-4f12-a4e3-6822833c44db.png)

Then it throw exception, like so:
{
    "errorMessage": `"Workflow faulted at 2021-09-17T02:33:50Z with error: The value \"\" is not of type \"System.Dynamic.ExpandoObject\" and cannot be used in this generic collection. (Parameter 'value')"`,
    "workflow": {
        "name": null,
        "version": 1,
        "instanceId": "06dc82c89f9d4cb7a6fc0d78d96833ed"
    }
}

**Comment by sfmskywalker** [comment]
OK. What happens when you JSON.stringify your object and use the resulting JSON string as the content?

**Comment by 4kb-wgh** [comment]
Thanks for your reply.

Interface code:
![image](https://user-images.githubusercontent.com/63579509/133749395-5e015230-80d0-4b3b-afaa-0e4589144c78.png)

SendHttpWorkflow:
![image](https://user-images.githubusercontent.com/63579509/133749262-de4b7468-3185-4d59-8735-ba1344fc7762.png)

I think it's caused by parsing objecj[2], but i don't know how to fixed it.

**Comment by sfmskywalker** [comment]
So the issue is not with sending an HTTP request, but with writing an HTTP response using WriteHttpResponse?

**Comment by 4kb-wgh** [comment]
Yes, interface return `object[]`, like so:
![image](https://user-images.githubusercontent.com/63579509/133767415-b5a0be28-a279-4553-92d3-09af66433f2e.png)

Somethings wrong with using WriteHttpResponse. I think it's about with `ExpandoObject`;

**Comment by sfmskywalker** [comment]
Can you send me your (stripped own) project so I can run it for myself?

**Comment by 4kb-wgh** [comment]
Yeah i have sent my test project to your email. Thanks

**Comment by sfmskywalker** [comment]
Thanks, I received t and the problem is now clear to me.

Here's what I will do:

The `SendHttpRequest` activity should allow you to configure in what shape or format you wish to receive the response. Right now, this is always `ExpandoObject` when the response is `application/json`, but in your particular case that is unsuitable, even more so because `"[\"\", null]"` is unparsable into an `ExpandoObject`.

So instead, you may want to just receive the response as a regular string value, or perhaps specify an actual .NET type like `System.Array<string>, System` (I'm guessing the .NET type name for a string array here, so it might not be exactly correct).

To allow for this, I will update the `SendHttpRequest` activity with a few more settings:

- The formatter to use to parse the response.
- If no formatter is selected, the response will be available as a regular string.
- If a formatter is selected, e.g. ExpandoObject, the response will be available as an ExpandoObject.
- A new formatter will be added called `StronglyTyped` which will allow you to specify the assembly-qualified type name of the model into which you want to parse the string.
- You will also be able to implement a custom formatter that receives the string so that you can do your own parsing. For example, you might receive an XML response, which cannot be parsed by JSON serializers.

I will need a little bit of time to work on this, so in the meantime, if you are stuck but want to continue, you might consider implementing your own custom version of `SendHttpRequest` that handles the response in anyway you want.

**Comment by 4kb-wgh** [comment]
Perfect, thanks.

**Comment by 4kb-wgh** [comment]
Perfect, thanks.

**Comment by sfmskywalker** [comment]
I just pushed a change (which will be available as part of the 2.3 prerelease packages once the build completes). By default, the content parser will be set to 'Auto', which means it will simply return the JSON string by default.

It has a few other parsers that you can select from:

![image](https://user-images.githubusercontent.com/938393/133884567-75d90ad0-83e6-4de9-b546-d81a7b529489.png)

- The JSON parser will return the raw JSON string that was returned from your API. This is useful if you want to do e.g. `JSON.stringify` in your HTTP Response activity.
- The .NET Type parser will try to parse the received JSON into a .NET type that you specify using an assembly-qualified type name. If no type name is specified, it will default to `ExpandoObject`.
- The Expando Object parser will try to parse the JSON into an `ExpandoObject`. This only works for JSON responses that are objects or arrays of objects, but not for primitives or arrays of primitives (your example API returns an array of primitives, so that won't work with this parsers; you might want to use the JSON parser instead).
- The JToken parsers is similar to the Expando Object parser, but instead parses the JSON string into a JToken.
- The File parser simply reads the responses as an array of bytes.

Let me know how this works out for you or not.

**Comment by sfmskywalker** [comment]
Closing this as solved, but feel free to follow up as you please.
